### PR TITLE
Image type for mapfish 2.x old api users

### PR DESCRIPTION
**Pull request by MaxComse**
Some time ago i used leaflet print plugin and happy with them. But 3 mapfish remove image type layers. I've return it in old api compatible. It probably usable for someone.

### CODE REVIEW

File: core/src/main/java/org/mapfish/print/servlet/oldapi/OldAPIMapPrinterServlet.java [hunk 1]
```diff
@@ -223,18 +223,20 @@ public final void getFile(@PathVariable final String id,
      * @param baseUrl  the path to the webapp
      * @param jsonpVar if given the result is returned as a variable assignment
      * @param req      the http request
-     * @param resp     the http response
+     * @param appId    the app request
+     * @param resp     the http response
```
Comment (tsauerwein) [inline core/src/main/java/org/mapfish/print/servlet/oldapi/OldAPIMapPrinterServlet.java]: Is this change related to adding the `image` layer? If not it would be good to do a separate PR.

File: core/src/main/java/org/mapfish/print/map/image/ImageLayer.java [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+                       final ImageParam params,
+                       final MetricRegistry registry) {
+        super(executorService, styleSupplier, params);   
+        this.params = params;
+        this.styleSupplier = styleSupplier;
+        this.executorService = executorService;
+    }
+
+    /**
+      * <p>Renders a image as layer.</p>
```
Comment (tsauerwein) [inline core/src/main/java/org/mapfish/print/map/image/ImageLayer.java]: "Renders an image..."

File: core/src/main/java/org/mapfish/print/map/image/ImageLayer.java [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+         
+         /**
+         * The base URL for the image file.  Used for making request.
+         */
+         public String baseURL;
+         
+         /**
+          * The extent of the image.  Used for placing image on map.
+          */
+         public double[] extent;
```
Comment (tsauerwein) [inline core/src/main/java/org/mapfish/print/map/image/ImageLayer.java]: It would be good to have a basic validation of this parameter. For example similar to this one: https://github.com/mapfish/mapfish-print/blob/master/core/src/main/java/org/mapfish/print/map/tiled/osm/OsmLayerParam.java#L133-L136

File: core/src/main/java/org/mapfish/print/map/image/ImageLayer.java [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+            return this.baseURL;
+        }
+    }
+
+    @Override
+    protected BufferedImage loadImage(final MfClientHttpRequestFactory requestFactory, 
+              final MapfishMapContext transformer) throws Throwable {
+        final ImageParam layerParam = this.params;
+        final URI commonUri = new URI(layerParam.getBaseUrl());
+        final int maxYIndex = 3;
```
Comment (tsauerwein) [inline core/src/main/java/org/mapfish/print/map/image/ImageLayer.java]: What is the purpose of this variable?

File: core/src/main/java/org/mapfish/print/map/image/ImageLayer.java [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+        final Double extentMaxX = layerParam.extent[2];
+        final Double extentMaxY = layerParam.extent[maxYIndex];
+        
+        final Rectangle paintArea = transformer.getPaintArea();
+        
+        final ReferencedEnvelope envelope = transformer.getBounds().toReferencedEnvelope(paintArea, transformer.getDPI());
+        final CoordinateReferenceSystem mapProjection = envelope.getCoordinateReferenceSystem();
+        
+        
+        //File imageFile = File.createTempFile("downloadedimage", "png");
```
Comment (tsauerwein) [inline core/src/main/java/org/mapfish/print/map/image/ImageLayer.java]: Please remove debugging statements like this.

File: core/src/main/java/org/mapfish/print/map/image/ImageLayer.java [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+        Closer closer = Closer.create();
+        final BufferedImage bufferedImage = new BufferedImage(paintArea.width, paintArea.height, TYPE_INT_ARGB_PRE);
```
Comment (tsauerwein) [inline core/src/main/java/org/mapfish/print/map/image/ImageLayer.java]: Looking at this code, I am wondering if it would not be better to inherit from [AbstractSingleImageLayer](https://github.com/mapfish/mapfish-print/blob/master/core/src/main/java/org/mapfish/print/map/image/AbstractSingleImageLayer.java), similar to [WmsLayer](https://github.com/mapfish/mapfish-print/blob/master/core/src/main/java/org/mapfish/print/map/image/wms/WmsLayer.java). What do you think?

File: core/src/main/java/org/mapfish/print/map/style/json/JsonStyleParserHelper.java [hunk 7]
```diff
@@ -93,6 +93,8 @@
     static final String JSON_EXTERNAL_GRAPHIC = "externalGraphic";
     static final String JSON_GRAPHIC_NAME = "graphicName";
     static final String JSON_GRAPHIC_OPACITY = "graphicOpacity";
+    static final String JSON_GRAPHIC_Y_OFFSET = "graphicYOffset";
```
Comment (tsauerwein) [inline core/src/main/java/org/mapfish/print/map/style/json/JsonStyleParserHelper.java]: Is this change related to this PR?

File: core/src/main/java/org/mapfish/print/servlet/oldapi/OldAPIMapPrinterServlet.java (hunk 1 above)
Comment (MaxComse) [inline core/src/main/java/org/mapfish/print/servlet/oldapi/OldAPIMapPrinterServlet.java]: Yes, it is not namely for image layer. This is bug fix as well. I think somebody just forget add this argument for old api compabilites.

File: core/src/main/java/org/mapfish/print/map/style/json/JsonStyleParserHelper.java (hunk 7 above)
Comment (MaxComse) [inline core/src/main/java/org/mapfish/print/map/style/json/JsonStyleParserHelper.java]: No. This is other feature...

File: core/src/main/java/org/mapfish/print/map/image/ImageLayer.java (hunk 4 above)
Comment (MaxComse) [inline core/src/main/java/org/mapfish/print/map/image/ImageLayer.java]: I have spotted it on OverviewMapAttribute class. Just for prevent "magic number" issue of code validator. Is different way available?

**Comment by MaxComse** [comment]
Hi! All recommends are done except graphicYOffset. Should i remove them?

File: core/src/main/java/org/mapfish/print/map/image/ImageLayer.java (hunk 3 above)
Comment (MaxComse) [inline core/src/main/java/org/mapfish/print/map/image/ImageLayer.java]: Validation added.
